### PR TITLE
Introducing Learning for Storages

### LINKED ISSUE #457: Add DRL storage units strategy
The framework is still missing a DRL based bidding strategy for the storage units. It should be added based on the Energy Informatics publication strategy.

### PR DESCRIPTION
**Pull request by kim-mskw**
# Pull Request

## Related Issue
Closes #457 

## Description
This PR introduces a learning strategy for storages implemented in accordance with the published strategy by @nick-harder, where the storage decides on the direction and the price height. The convergence has been tested and looks good. Also the example_02e has the set of finetuned hyperparameter to enable good storage learning performance.

## Changes Proposed
- added a new DRL based bidding strategy for the storage unit
- prepared example 02e with a learning storage unit
- switched from using SOC in the storage unit as relative value to absolute value to avoid confusions
- enhanced the docstrings of both learning strategies for the power plant as well as the storage
- adjusted tests and bidding strategies to accomodate the above changes
- changed pyomo usage for dmas storages to align with usage in other moduls

## Testing
General usability as well as learning performance was tested. No multi-agent tests were done yet, this will come in the following PRs.

## Checklist
Please check all applicable items:

- [x] Code changes are sufficiently documented (docstrings, inline comments, `doc` folder updates)
- [x] New unit tests added for new features or bug fixes
- [x] Existing tests pass with the changes
- [x] Reinforcement learning examples are operational (for DRL-related changes)
- [x] Code tested with both local and Docker databases
- [x] Code follows project style guidelines and best practices
- [x] Changes are backwards compatible, or deprecation notices added
- [ ] New dependencies added to `pyproject.toml`
- [x] A note for the release notes `doc/release_notes.rst` of the upcoming release is included
- [x] Consent to release this PR's code under the GNU Affero General Public License v3.0

## Additional Notes (if applicable)
[Any additional information, concerns, or areas you want reviewers to focus on]

## Screenshots (if applicable)
@nick-harder 
image

@ModdyLP 
image

### CODE REVIEW

File: examples/examples.py [hunk 1]
```diff
@@ -100,8 +102,8 @@
     - local_db: without database and grafana
     - timescale: with database and grafana (note: you need docker installed)
     """
-    data_format = "local_db"  # "local_db" or "timescale"
-    example = "small"
+    data_format = "timescale"  # "local_db" or "timescale"
+    example = "small_learning_storage"
```
Comment (maurerle) [inline examples/examples.py:106]: please do not adjust the default values of the examples file

File: assume/strategies/learning_strategies.py [hunk 2]
```diff
@@ -506,3 +506,550 @@ def load_actor_params(self, load_path):
...
+        #technical parameters of power plant
+        self.max_power_charge = kwargs.get("max_power_charge", 100)
+        self.max_power_discharge = kwargs.get("max_power_discharge", 90)
+        self.efficiency_charge = kwargs.get("efficiency_charge", 0.95)
+        self.efficiency_discharge = kwargs.get("efficiency_discharge", 0.95)
+        self.min_volume = kwargs.get("min_volume", 1)
+        self.max_volume = kwargs.get("max_volume", 190)
+        self.variable_cost_charge = kwargs.get("variable_cost_charge", 30)
+        self.variable_cost_discharge = kwargs.get("variable_cost_discharge", 30)
+        self.natural_inflow = kwargs.get("natural_inflow", 0)
```
Comment (maurerle) [inline assume/strategies/learning_strategies.py:531]: all these seem to not be used at all..?

**Review by maurerle — COMMENTED** [review]
Would be very good to have a test for the initialization of the Bidding Strategies.

I think it is fine to include both, as long as there is also code which somehow tests the functionality of it.

File: examples/examples.py (hunk 1 above)
Comment (kim-mskw) [reply examples/examples.py:106]: Nick wanted to continue working on both startegies, so I pushed the example I used, but yes indeed it should be changed back!

**Comment by nick-harder** [comment]
I have now adjusted the code and the strategy to bring it in line with the publication. Also I have fine tuned the hyperparameters to fit the performance. Everything looks good now. I have not tested the multi-agent setting, but this will be done during the next pool request, where I will also prepare the data I have used for the paper and run the same simulations to compare results. Attached is a screenshot of the evaluations for the learning

image

**Comment by nick-harder** [comment]
@kim-mskw @maurerle this is ready so far. it would be great if you could take a look at it

File: assume/common/base.py [hunk 3]
```diff
@@ -807,6 +812,38 @@ def __init__(
         # them into suitable format for recurrent neural networks
         self.num_timeseries_obs_dim = num_timeseries_obs_dim
 
+    def load_actor_params(self, load_path):
```
Comment (maurerle) [inline assume/common/base.py:815]: I would rather move the whole LearningStrategy from base to learning_strategies.py instead of adding to much here..

**Review by maurerle — APPROVED** [review]

File: assume/common/base.py (hunk 3 above)
Comment (nick-harder) [reply assume/common/base.py:815]: @maurerle I believe not all DRL based bidding strategies will be always contained within a single file. For now we already have learning_strategies.py and learning_advanced_startegies.py. So your suggestions would not work. But overall it is a good idea to maybe move it somewhere else from the base, but where...

**Comment by kim-mskw** [comment]
Thanks @nick-harder! I am just missing one sentence about the reasoning behind the defintion of ignored bids somewhere.

**Pull request merged**